src/mongocxx/instance.cpp

```diff
@@ -45,5 +82,11 @@ instance::instance(instance &&) noexcept = default;
 instance &instance::operator=(instance &&) noexcept = default;
 instance::~instance() = default;
 
+void instance::set_logger(std::unique_ptr<logger> logger) {
+    std::lock_guard<std::mutex> lk(_impl->logger_mtx);
+    _impl->user_logger = std::move(logger);
```

Comment:
This is unfortunately still racy. Thread A could be the currently configured logger while thread B is calling into set_logger.

Per discussion, lets make the logger a ctor arg to instance, and remove set_logger (and the mutex).